File: 103-binary-tree-zigzag-level-order-traversal/binary-tree-zigzag-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        queue<TreeNode*> q;
        vector<vector<int>> lis;
        if(root==NULL)return lis;
        q.push(root);
        bool flag=false;
        while(!q.empty()){
            int n=q.size();
            flag= !flag;
            vector<int> tmplis;
            for(int i=0;i<n;i++){
                TreeNode* tmp=q.front();
                q.pop();
                if(tmp->left!=NULL)q.push(tmp->left);
                if(tmp->right!=NULL)q.push(tmp->right);
                if(flag) tmplis.push_back(tmp->val);
                else tmplis.insert(tmplis.begin(),tmp->val);
            }
            lis.push_back(tmplis);
        }
        return lis;
    }
};

File: 103-binary-tree-zigzag-level-order-traversal/binary-tree-zigzag-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        lis= list()
        if not root: return lis
        q=deque()
        q.append(root)
        flag=False
        while q:
            n=len(q)
            flag=not flag
            tmplis = list()
            for i in range(n):
                temp=q.popleft()
                if temp.left: q.append(temp.left)
                if temp.right: q.append(temp.right)
                if flag:
                    tmplis.append(temp.val)
                else: tmplis.insert(0,temp.val)
            lis.append(tmplis)
        return lis



        